main/src/main/scala/sbt/internal/Load.scala

```diff
@@ -943,142 +943,160 @@ private[sbt] object Load {
-            }
-          val newProjects = moreProjects ++ projectLevelExtra
-          val newAcc = finalRoot +: (acc ++ otherProjects.projects)
-          val newGenerated =
-            generated ++ otherProjects.generatedConfigClassFiles ++ generatedConfigClassFiles
-          load(newProjects, newAcc, newGenerated, finalRoot.commonSettings)
-        case Nil =>
-          val projectIds = acc.map(_.id).mkString("(", ", ", ")")
-          log.debug(s"[Loading] Done in $buildBase, returning: $projectIds")
-          LoadedProjects(acc, generatedConfigClassFiles)
+      // TODO: We assume here the project defined in a build.sbt WINS because the original was a
+      // phony.  However, we may want to 'merge' the two, or only do this if the original was a
+      // default generated project.
+      val root = rootOpt.getOrElse(p)
+      val (finalRoot, projectLevelExtra) = finalizeProject(root, files, extraFiles, true)
+      val newProjects = rest ++ discovered ++ projectLevelExtra
+      val newAcc = acc :+ finalRoot
+      val newGenerated = generated ++ generatedConfigClassFiles
+      loadTransitive1(newProjects, newAcc, newGenerated, finalRoot.commonSettings)
     }
 
+    // Load all config files AND finalize the project at the root directory, if it exists.
+    // Continue loading if we find any more.
+    newProjects match
+      case Seq(next, rest @ _*) =>
+        log.debug(s"[Loading] Loading project ${next.id} @ ${next.base}")
+        discoverAndLoad(next, rest)
+      case Nil if makeOrDiscoverRoot =>
+        log.debug(s"[Loading] Scanning directory $buildBase")
+        val DiscoveredProjects(rootOpt, discovered, files, extraFiles, generated) = discover(
+          buildBase
+        )
+        val discoveredIdsStr = discovered.map(_.id).mkString(",")
+        val (root, expand, moreProjects, otherProjects) =
+          rootOpt match
+            case Some(root) =>
+              log.debug(s"[Loading] Found root project ${root.id} w/ remaining $discoveredIdsStr")
+              (root, true, discovered, LoadedProjects(Nil, Nil))
+            case None =>
+              log.debug(s"[Loading] Found non-root projects $discoveredIdsStr")
+              // Here we do something interesting... We need to create an aggregate root project
+              val root = {
+                val defaultID = autoID(buildBase, context, Nil)
+                if discovered.isEmpty || java.lang.Boolean.getBoolean("sbt.root.ivyplugin") then
+                  BuildDef.defaultProject(defaultID, buildBase)
+                else BuildDef.generatedRootSkipPublish(defaultID, buildBase, Nil)
+              }
+              val otherProjects =
+                loadTransitive1(
+                  newProjects = discovered,
+                  acc = acc,
+                  generated = Nil,
+                  commonSettings0 = commonSettings
+                    ++ expandCommonSettingsPerBase1(buildBase.getCanonicalFile()),
```

Comment:
The key change is here. Given `hello/build.sbt` and `foo` subproject, it will now load common settings from `hello/build.sbt` as expected not just `hello/foo/build.sbt`.